src/backend/starknet.cairo

```diff
@@ -295,4 +302,53 @@ namespace Internals {
 
         return _save_storage(starknet_address, storage_start + DictAccess.SIZE, storage_end);
     }
+
+    func _save_valid_jumpdests{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        starknet_address: felt, dict_start: DictAccess*, dict_end: DictAccess*
+    ) {
+        alloc_locals;
+        let (local keys_start: felt*) = alloc();
+        let dict_len = dict_end - dict_start;
+        let (local len, _) = unsigned_div_rem(dict_len, DictAccess.SIZE);
+        local range_check_ptr = range_check_ptr;
+
+        if (dict_len == 0) {
+            return ();
+        }
+
+        tempvar keys_len = 0;
+        tempvar keys = keys_start;
+        tempvar len = len;
+        tempvar dict = dict_start;
+
+        loop:
+        let keys_len = [ap - 4];
+        let keys = cast([ap - 3], felt*);
+        let len = [ap - 2];
+        let dict = cast([ap - 1], DictAccess*);
+        let is_valid = dict.new_value;
+
+        if (is_valid != 0) {
+            assert [keys] = dict.key;
+            tempvar keys_len = keys_len + 1;
+            tempvar keys = keys + 1;
+            tempvar len = len - 1;
+            tempvar dict = dict + DictAccess.SIZE;
+        } else {
+            tempvar keys_len = keys_len;
+            tempvar keys = keys;
+            tempvar len = len - 1;
+            tempvar dict = dict + DictAccess.SIZE;
+        }
```

Comment:
can you move the common part out of the `if/else`? `jmp if != 0` doesn't update `[ap]` so it should still work as is. If not, maybe use raw `jmp` instead

Reply:
I don't understand how I can move the common part (dict = dict + DictAccess.SIZE) outside of the if/else. It just won't compile.